src/geff/testing/data.py

```diff
@@ -0,0 +1,524 @@
+        axis_units_list.append("nm")
+    if include_x:
+        axis_names_list.append("x")
+        axis_units_list.append("nm")
+
+    axis_names: tuple[Axes, ...] = cast("tuple[Axes, ...]", tuple(axis_names_list))
+    axis_units = tuple(axis_units_list)
+
+    # Generate nodes with flexible count
+    if node_id_dtype == "str":
+        # For string dtype, create string representations of numbers
+        nodes = np.array([f"node_{i}" for i in range(num_nodes)], dtype=node_id_dtype)
+    else:
+        # For numeric dtypes, use arange
+        nodes = np.arange(num_nodes, dtype=node_id_dtype)
+
+    # Generate spatiotemporal coordinates with flexible dimensions
+    t = (
+        np.array(
+            [(i * 5 // num_nodes) + 1 for i in range(num_nodes)],
+            dtype=node_axis_dtypes["time"],
+        )
+        if include_t
+        else np.array([], dtype="float64")  # Default dtype when time not included
+    )
+    z = (
+        np.linspace(0.5, 0.1, num_nodes, dtype=node_axis_dtypes["position"])
+        if include_z
+        else np.array([], dtype="float64")  # Default dtype when position not included
+    )
+    y = (
+        np.linspace(100.0, 500.0, num_nodes, dtype=node_axis_dtypes["position"])
+        if include_y
+        else np.array([], dtype="float64")  # Default dtype when position not included
+    )
+    x = (
+        np.linspace(1.0, 0.1, num_nodes, dtype=node_axis_dtypes["position"])
+        if include_x
+        else np.array([], dtype="float64")  # Default dtype when position not included
+    )
+
+    # Generate edges with flexible count (ensure we don't exceed possible edges)
+    max_possible_edges = (
+        num_nodes * (num_nodes - 1) // 2 if not directed else num_nodes * (num_nodes - 1)
+    )
+    actual_num_edges = min(num_edges, max_possible_edges)
+
+    # Create edges ensuring we don't create duplicates
+    edges: list[list[Any]] = []
+    edge_count = 0
+
+    # For undirected graphs, we need to be more careful about duplicates
+    if not directed:
+        # Create a simple chain first, then add cross edges
+        for i in range(min(actual_num_edges, num_nodes - 1)):
+            source_idx = i
+            target_idx = i + 1
+            if node_id_dtype == "str":
+                edges.append([f"node_{source_idx}", f"node_{target_idx}"])
+            else:
+                edges.append([int(source_idx), int(target_idx)])
+            edge_count += 1
+
+        # Add remaining edges as cross connections
+        remaining_edges = actual_num_edges - edge_count
+        for i in range(remaining_edges):
+            source_idx = i % (num_nodes - 2)
+            target_idx = (i + 2) % (num_nodes - 1) + 1
+            if source_idx != target_idx:
+                if node_id_dtype == "str":
+                    edges.append([f"node_{source_idx}", f"node_{target_idx}"])
+                else:
+                    edges.append([int(source_idx), int(target_idx)])
```

Comment:
We are casting to int here. I think this is ok because we're only allowing int, but we should verify.